Snap 889

vibhaska opened the pull request:
## Changes proposed in this pull request
Fixes race between GII and an update going in parallel (As mentioned in point 2 below).

The scenario of defect is mentioned as below:
Primary Bucket:

1. Gets update operation:
storegemfire1_5242/system-01-01.log : 
17/10/12 09:42:11.663 IST Function Execution Processor12<tid=0x2b4> INFO storegemfire1_localhost_5242: QueryDistribution: GS#prepMinion: preparing update trade.securities set exchange =? where sec_id = ? and tid = ?

2. Sends to Secondary, secondary enqueues the delta. But, at the end of GII it is not applied to the oldValue as seen from the log. We think the delta value corresponding to update came after requestForDelta is done. There is race condition between update and GII.
The node goes down after 25 seconds.


storegemfire1_21727/system-01-01.log
17/10/12 09:42:11.646 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x1a8> INFO storegemfire1_localhost_21727: Region _B__TRADE_SECURITIES_110 requesting initial image from 192.168.1.164(storegemfire1_localhost_5242:5242)<v21>:6902

17/10/12 09:42:11.669 IST P2P message reader for 192.168.1.164(storegemfire1_localhost_5242:5242)<v21>:6902 unshared ordered uid=32 dom #2<tid=0x11c> INFO storegemfire1_localhost_21727: enqueDelta: new list of deltas with delta: SerializableDelta@0x63581f25: changedRow= col[0]=null col[1]=null col[2]=null col[3]=tse col[4]=null, validColumns=(3)for re =VersionedBucketRowLocationThinDiskRegionEntryHeap@8cb6d38(key=(825); rawValue=(2,0,0,3,57,112,119,122,109,118,53,20,1,65,54,-6,-114,59,-102,-19,-122,-96,102,115,101,0,0,0,5,5,11,22,25); lockState=0x0; bucketId=110) oldVal = [B@429fb399



17/10/12 09:42:11.690 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x1a8> INFO storegemfire1_localhost_21727: recording rv0 for 1db6baf6-3d8e-4c62-b27c-1943280b3713 rvv {7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv18 bsv1 bs=[], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv1 bs=[]} contains true

17/10/12 09:42:11.695 IST P2P message reader for 192.168.1.164(storegemfire1_localhost_5242:5242)<v21>:6902 unshared ordered uid=32 dom #2<tid=0x11c> INFO storegemfire1_localhost_21727: recording rv19 for 7e630fd5-4c7a-4f66-8542-1bf92ce1dda4 rvv {7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 bsv18 bs=[0,1], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv1 bs=[]} contains true


3. Secondary restarts: comes up with rv 18. It missed the ListOfDelta value related to the updated entry. It comes up with oldValue and earlier regionversion.
17/10/12 09:42:42.123 IST vm_3_thr_18_store1_localhost_21727<tid=0x303> INFO storegemfire1_localhost_21727:
17/10/12 09:42:42.123 IST vm_3_thr_18_store1_localhost_21727<tid=0x303> INFO storegemfire1_localhost_21727:
17/10/12 09:43:03.898 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4cf> INFO storegemfire1_localhost_21727: Initializing region _B__TRADE_SECURITIES_110
17/10/12 09:43:03.898 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4cf> INFO storegemfire1_localhost_21727: recording rv28 for 1db6baf6-3d8e-4c62-b27c-1943280b3713 rvv {7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv18 bsv18 bs=[0], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv1 bs=[]} contains true


4. Primary restarts, comes up with rv 19, concludes the rv19 to be unfinished operation and rejects the updated entry. After GII, both primary and secondary have rv18 as latest region version.
17/10/12 09:42:42.126 IST vm_6_thr_21_store1_localhost_5242<tid=0x33e> INFO storegemfire1_localhost_5242:

17/10/12 09:43:03.973 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: recording rv19 for 7e630fd5-4c7a-4f66-8542-1bf92ce1dda4 rvv {7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 bsv1 bs=[], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv1 bs=[]} contains true


17/10/12 09:43:03.973 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: Region:/__PR/_B__TRADE_SECURITIES_110 found unfinished operation key=tableinfo(true).CompactCompositeRegionKey@40e6f85e=(825),member=7e630fd5-4c7a-4f66-8542-1bf92ce1dda4,region version=19


17/10/12 09:43:03.978 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: Region /__PR/_B__TRADE_SECURITIES_110 recovered with EndGII flag, rvv is RegionVersionVector[7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 gc0 localVersion=19 local exceptions=[]} others={7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 bsv1 bs=[], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv28 bs=[0]}, gc={}]. recovered rvv is RegionVersionVector[7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 gc0 localVersion=19 local exceptions=[]} others={7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 bsv1 bs=[], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv28 bs=[0]}, gc={}]. Do delta GII

17/10/12 09:43:04.000 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: Processing unfinished operation:entry=VersionedBucketRowLocationThinDiskRegionEntryHeap@3a3102f6(key=(825); rawValue=(2,0,0,3,57,112,119,122,109,118,53,20,1,65,54,-6,-114,59,-102,-19,-122,-96,102,115,101,0,0,0,5,5,11,22,25); lockState=0x0; bucketId=110)
17/10/12 09:43:04.000 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: _B__TRADE_SECURITIES_110 is done getting image from 192.168.1.164(storegemfire1_localhost_21727:21727)<v28>:9371. isDeltaGII is true rvv is RegionVersionVector{rv19 gc0}
17/10/12 09:43:04.000 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: Region _B__TRADE_SECURITIES_110 initialized persistent id: /192.168.1.164:/home/vivek/work/bugs/1689/hydra/hydraLogs/sqlDisk/2017-10-11-19-56-36/randomPartitionReplPersistTablesHA-1012-092757/vm_6_store1_disk/persistSecu created at timestamp 1507781569942 version 0 diskStoreId 7e630fd5-4c7a-4f66-8542-1bf92ce1dda4 name null with data from 192.168.1.164(storegemfire1_localhost_21727:21727)<v28>:9371.
17/10/12 09:43:04.000 IST Recovery thread for bucket _B__TRADE_SECURITIES_110<tid=0x4db> INFO storegemfire1_localhost_5242: recording rv0 for 7e630fd5-4c7a-4f66-8542-1bf92ce1dda4 rvv {7e630fd5-4c7a-4f66-8542-1bf92ce1dda4={rv19 bsv1 bs=[][e(n=20 p=18; rb=[])], 1db6baf6-3d8e-4c62-b27c-1943280b3713={rv28 bsv1 bs=[]} contains true
 

## Patch testing
Hydra test
Precheckin

## ReleaseNotes changes
No

## Other PRs 
No

Commit:
Take care of NPE coming while writing message in response of any earlier exception.

Commit:
Fix from Suranjan to fix race between parallel update and GII.

kneeraj commented:
Trying to explain the changes in a nutshell.

There was a race possible between a region getting initialized and an on the fly update ( coming as delta ) when the region is uninitialized. So for these deltas, becuase of the race there was no chance of fetching the base value explicitly which we do at the end of the GII.

@sumwale please review as we three ( @suranjan @vivekwiz and I ) have discussed the changes.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java:
```diff
@@ -3649,8 +3652,28 @@ protected final void releaseAfterGetInitialImageLatch() {
+    if (deltaLock.isWriteLocked()) {
+      deltaLock.writeLock().unlock();
+    }
+  }
```
The isWriteLocked() checks have race conditions. It also will cause other threads to skip locking if already locked by another thread. Is that the required behaviour?? To me that defeats the purpose of these locks which were meant to guard multiple threads trying to enqueue concurrently (or is that the purpose?).

kneeraj replied:
>> The isWriteLocked() checks have race conditions.

Since the write lock is attempted only by the gii processor thread so for write locks there isn't any race.

sumwale replied:
So why is this check there at all? Is it being expected that write lock may not have been taken at this point? As I mentioned that such a case would be a bug and so this check should be removed.

kneeraj replied:
I think since the releaseAfterGetInitialImageLatch() will be called both in case of success/failure(exception somewhere in between gii process) so this check would have been required, otherwise blindly releasing it was fine.
BTW @suranjan we can perhaps have a read write lock just for delta scenario and write lock/unlock around requestUnappliedDeltas method and instead of region init flag, check delta fetched flag or something in the basic putter thread.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java:
```diff
@@ -3649,8 +3652,28 @@ protected final void releaseAfterGetInitialImageLatch() {
     }
     releaseLatch(this.initializationLatchAfterGetInitialImage);
+
+    writeUnlockEnqueueDelta();
```
How are we sure that this route has always acquired the lock before? And shouldn't this be in some finally block by the caller rather than being here.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DistributedRegion.java:
```diff
@@ -1620,6 +1620,7 @@ protected void cleanUpAfterFailedGII(boolean recoverFromDisk) {
     } finally {
       try {
         doPostEntryDestructionCleanup(true, null, false, giiIndexLockAcquired);
+        writeUnlockEnqueueDelta();
```
Where is the corresponding lock call? Are we sure that this will be always locked here? Also this should go into the finally block below.

sumwale reviewed (commented):
Not sure I understand what exactly is the race here. Can you explain what is the race that this read-write lock is trying to fix in simple steps (without log references) and how this RW lock fixes that?

Secondly there are issues with the RW locks locking/unlocking as mentioned in comments. Also am very skeptical about the lock-unlock matching each other in all cases. The basic problem I see is the "isWriteLocked()" workaround due to lack of matching I think in some cases but that defeats the purpose of the write lock as far as I can tell.

IMO the clean way to fix all these delta related issues on secondary is to get rid of delta sends from source itself. The savings, especially for row tables, are small while the complications large. I had done the same for column deltas knowing the various issues with delta handling and if it goes wrong for column tables then a whole bunch of columns can go wrong (see EntryEventImpl.applyDelta the check for SeriailzedDiskBuffer). So I think better option is to null the delta for all cases in EntryEventImpl.applyDelta and we can revisit this later for large column deltas in future.

kneeraj commented:
>> Not sure I understand what exactly is the race here. Can you explain what is the race that this read-write lock is trying to fix in simple steps (without log references) and how this RW lock fixes that?

@sumwale we saw the race here:
Let's say a bucket region is doing the gii. Till the gii is completed the region is uninitialized.
Now updates in the form of delta can keep on coming. There can be one or more update which can land here just after the gii process is almost complete but is still to toggle the initialized flag. So let's say the gii requestor thread has processed the image it got from source. Then even the message of 'requestUnappliedDeltas' has been sent and applied. Now between this time and the initialized being set to true if an update comes, it will still see the region as uninitialized and create a list of delta object. And then the gii thread goes ahead and sets the region as initialized, without the knowledge that more ListOfDelta object got created.
So the base value on which the delta was needed to be applied will never come and these kind of entries will remain in the map.
The first explanation which I gave was probably too short. Hope this makes things clearer.

kneeraj commented:
>> Secondly there are issues with the RW locks locking/unlocking as mentioned in comments.

The write lock will be taken only by the gii requestor thread. Please see the requestUnappliedDeltas method which is called as part of the gii at the end. It will be released after the gii is complete and the region is marked as initialized. Now the update processing thread checks the region is initialized or not twice ( once more after taking the read lock ) just to avoid the race in locks which you are suspecting. Later the normal puts will not be affected as the read lock is tried only when the region is uninitialized.

sumwale commented:
Ok, got the scenario. Won't it be better to set a flag on ImageState instead? Before the call to requestUnappliedDeltas set a flag, and then the incoming puts will check that flag instead of owner.initiailzed (can see the checks in AbstractRegionMap.enqueDelta, HDFSRegionMapDelegate.getValueDuringGII). One would need to make ImageState.addDeltaEntry as returning a boolean to add only if that flag is not set atomically and only then rest of method should follow. Perhaps that call should go inside listOfDeltasCreator.

Quite apprehensive about the cross-module read-write lock approach. Still can see paths where lock/unlock can mismatch.

kneeraj commented:
Yes. No need to check if the entire region is initialized or not. I think my above recommendation in one of the comment:
>> BTW @suranjan we can perhaps have a read write lock just for delta scenario and write lock/unlock around requestUnappliedDeltas method and instead of region init flag, check delta fetched flag or something in the basic putter thread.

perhaps is similar to what you are suggesting. @vivekwiz  and @suranjan We should check the approach suggested.

kneeraj commented:
I think sending the full value from primary to secondary copy at least as you have suggested looks to be a good option and more risk free.

vibhaska commented:
On alternate approach, I sent an email with exceptions I am getting.

sumwale commented:
> On alternate approach, I sent an email with exceptions I am getting.

The only reason I can think why exceptions might be generated is if the operation type on secondary is being mistaken as a create rather than a put. Might need to check event type in EntryEventImpl (as seen by GfxdIndexManager) and change it if required.

vibhaska commented:
@sumwale Looking at the code, yes isPutDML flag is false; and its case of Update. Should this be changed?

Commit:
Merge branch 'snappy/master' into SNAP-889

Commit:
Used flag to check for unapplied delta message

  Instead of waiting for region getting initialized, wait
  till unapplied deltas are applied to continue with basicput

  Use read write lock to make sure that when unapplied deltas are
  being fetched and applied we don't do any delta put

suranjan commented:
@sumwale @kneeraj Please review the changes once again.

Commit:
Merge branch 'snappy/master' into SNAP-889

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/AbstractRegionMap.java:
```diff
@@ -3864,12 +3864,17 @@ public RegionEntry basicPut(EntryEventImpl event,
       // wait for the initialization to get over which will not finish
       // until this testLatch is counted down from the test code
       if (DistributedRegion.testLatch == null) {
```
We can still keep the !owner.initialized() condition as acquiring the readLock is not necessary when the region is initialized.

suranjan replied:
There is a check just above for the same. So, if region is initialized, this code path will not be executed.

kneeraj replied:
Cool. The context was not that wide to include that statement so missed it.

kneeraj reviewed (commented):
LGTM except a comment to not acquire read lock when the region is initialized.

Commit:
Merge branch 'snappy/master' into SNAP-889
